Sweetener.Logging/LogPattern.cs

```diff
@@ -0,0 +1,385 @@
+                // Parse the format between the comma ',' and either the colon ':' or closed curly brace '}'
+                if (pattern[i] == ',')
+                {
+                    int alignmentStart = ++i;
+                    for (; i < pattern.Length && pattern[i] != ':' && pattern[i] != '}'; i++)
+                    { }
+
+                    if (i >= pattern.Length)
+                        throw new FormatException("Input string was not in a correct format");
+
+                    return pattern.Substring(alignmentStart, i - alignmentStart);
+                }
+
+                return null;
+            }
+
+            private static string ParseFormat(ref int i, string pattern)
+            {
+                // Parse the format between the colon ':' and the closed curly brace '}'
+                if (pattern[i] == ':')
+                {
+                    int formatStart = ++i;
+                    for (; i < pattern.Length && pattern[i] != '}'; i++)
+                    { }
+
+                    if (i >= pattern.Length)
+                        throw new FormatException("Input string was not in a correct format");
+
+                    return pattern.Substring(formatStart, i - formatStart);
+                }
+
+                return null;
+            }
+
+            private static Parameter ParseParameter(ref int i, string pattern)
+            {
+                // Need to keep track of the number of characters read because it is not
+                // necessarily the same as the size of the name. Spaces are allowed before or after.
+                int nameStart = -1, nameEnd = -1;
+                for (; i < pattern.Length; i++)
+                {
+                    char c = pattern[i];
+                    // There are currently no patterns with numbers enabled
+                    if (char.IsLetter(c) && nameEnd == -1)
+                    {
+                        if (nameStart == -1)
+                            nameStart = i;
+                    }
+                    else if (c == ' ')
+                    {
+                        if (nameStart != -1 && nameEnd == -1)
+                            nameEnd = i;
+                    }
+                    else if (c == ',' || c == ':' || c == '}')
+                    {
+                        if (nameEnd == -1)
+                            nameEnd = i;
+
+                        break;
+                    }
+                    else
+                    {
+                        throw new FormatException("Input string was not in a correct format");
+                    }
+                }
+
+                // Did we reach the end without breaking?
+                if (i >= pattern.Length)
+                    throw new FormatException("Input string was not in a correct format");
+
+                string name = pattern.Substring(nameStart, nameEnd - nameStart);
+
+                // TODO: Case insensitive?
+                switch (name)
```

Comment:
Any other thoughts on supported parameters out-of-the-box right now? For inspiration, here are the [patterns](https://logging.apache.org/log4j/2.x/manual/layouts.html#Patterns) for Log4J

Reply:
The calling class? Not all logged messages will be stack traces and could be useful to quickly see where the log statement came from.

Reply:
Yeah location information can be really helpful. It can help drill down into the method instantly with info like Class Name, Method Name, Line Number, etc. I think Log4J calls this the location info. But the caveat is that deriving location info at runtime is expensive. So how can we expose this info?

1. People are aware that these parameters are expensive and may slow down their programs. We document this in the wiki or something.
2. We expect users to include this information explicitly themselves in their messages. It is available at building time and users should just specify it manually.
3. We can perhaps do something a bit trickier, where we include an msbuild target. If a user indicates they want this information, we scan the code for these log calls and add this metadata at build time.